text/0000-process-stdio-redirection.md

```diff
@@ -0,0 +1,126 @@
+- Feature Name: process-stdio-redirection
+- Start Date: 2015-04-10
+- RFC PR:
+- Rust Issue:
+
+# Summary
+
+Update the `std::process` API with the ability to redirect stdio of child
+processes to any opened file handle or equivalent.
+
+# Motivation
+
+The current API in `std::process` allows to either pipe stdio between parent and
+child process or redirect stdio to `/dev/null`. It would also be largely useful
+to allow stdio redirection to any currently opened file or pipe handles. This
+would allow redirecting stdio with a physical file or even another process (via
+OS pipe) without forcing the parent process to buffer the data itself.
+
+For example, one may wish to spawn a process which prints gigabytes
+of data (e.g. logs) and use another process to filter through it, and save the
+result to a file. The current API would force the parent process to dedicate a
+thread just to buffer all data between child processes and disk, which is
+impractical given that the OS can stream the data for us via pipes and file
+redirection.
+
+# Detailed design
+
+First, the standard library should provide an OS agnostic way of creating OS
+in-memory pipes, i.e. a `Pipe`, providing reader and writer handles as
+`PipeReader` and `PipeWriter` respectively. The reader and writer will simply be
+an abstraction over the OS specific file descriptor/HANDLE, and will implement
+the respective `Read`/`Write` trait, making it impossible to confuse the two
+together. In addition, each should implement the appropriate `AsRaw{Fd,
+Handle}`/`FromRaw{Fd, Handle}` for wrapping/unwrapping of the OS handles.
+
+On Unix systems the pipe can be created with `libc::pipe` and the file
+descriptors wrapped as appropriate (e.g. `sys::fs2::File`). On Windows the pipe
+can be created via Windows' `CreatePipe` (a stub for which is missing in
+`liblibc` at the moment) and the resulting HANDLEs also appropriately wrapped
+(using `sys::fs2::File`).
+
+This proposal considers making the `Pipe`'s reader and writer fields public
+to allow easily moving ownership of the two handles, but any other appropriate
+interface is acceptable.
+
+```rust
+pub struct Pipe {
+	pub reader: PipeReader,
+	pub writer: PipeWriter,
+}
+
+pub struct PipeReader(sys::fs2::File);
```

Comment:
Actually I completely failed to consider the possibility of extrancting the fd/HANDLE from `ChildStd{in,out,err}` (and my apologies for misunderstanding one of your earlier comments). It should work well for hooking up any number of processes together (in most cases) without having to create explicit pipes! (though it might still be nice to have a public interface for creating them)

 I'll rework the proposal to include this strategy over the next few days!